src/core/pdf_manager.js

```diff
@@ -15,152 +15,149 @@
+    return this.pdfDocument.cleanup();
+  }
 
-    ensure: function BasePdfManager_ensure(obj, prop, args) {
-      unreachable('Abstract method `ensure` called');
-    },
+  ensure(obj, prop, args) {
+    unreachable('Abstract method `ensure` called');
+  }
 
-    requestRange: function BasePdfManager_requestRange(begin, end) {
-      unreachable('Abstract method `requestRange` called');
-    },
+  requestRange(begin, end) {
+    unreachable('Abstract method `requestRange` called');
+  }
 
-    requestLoadedStream: function BasePdfManager_requestLoadedStream() {
-      unreachable('Abstract method `requestLoadedStream` called');
-    },
+  requestLoadedStream() {
+    unreachable('Abstract method `requestLoadedStream` called');
+  }
 
-    sendProgressiveData: function BasePdfManager_sendProgressiveData(chunk) {
-      unreachable('Abstract method `sendProgressiveData` called');
-    },
+  sendProgressiveData(chunk) {
+    unreachable('Abstract method `sendProgressiveData` called');
+  }
 
-    updatePassword: function BasePdfManager_updatePassword(password) {
-      this._password = password;
-    },
+  updatePassword(password) {
+    this._password = password;
+  }
 
-    terminate: function BasePdfManager_terminate() {
-      unreachable('Abstract method `terminate` called');
-    },
-  };
+  terminate() {
+    unreachable('Abstract method `terminate` called');
+  }
+}
 
-  return BasePdfManager;
-})();
+class LocalPdfManager extends BasePdfManager {
+  constructor(docId, data, password, evaluatorOptions, docBaseUrl) {
+    super();
 
-var LocalPdfManager = (function LocalPdfManagerClosure() {
-  function LocalPdfManager(docId, data, password, evaluatorOptions,
-                           docBaseUrl) {
     this._docId = docId;
     this._password = password;
     this._docBaseUrl = docBaseUrl;
     this.evaluatorOptions = evaluatorOptions;
-    var stream = new Stream(data);
+
+    const stream = new Stream(data);
     this.pdfDocument = new PDFDocument(this, stream);
     this._loadedStreamCapability = createPromiseCapability();
     this._loadedStreamCapability.resolve(stream);
   }
 
-  Util.inherit(LocalPdfManager, BasePdfManager, {
-    ensure: function LocalPdfManager_ensure(obj, prop, args) {
-      return new Promise(function (resolve, reject) {
-        try {
-          var value = obj[prop];
-          var result;
-          if (typeof value === 'function') {
-            result = value.apply(obj, args);
-          } else {
-            result = value;
-          }
-          resolve(result);
-        } catch (e) {
-          reject(e);
+  ensure(obj, prop, args) {
+    return new Promise(function(resolve, reject) {
+      try {
+        const value = obj[prop];
+        let result;
+        if (typeof value === 'function') {
+          result = value.apply(obj, args);
+        } else {
+          result = value;
         }
-      });
-    },
+        resolve(result);
+      } catch (e) {
+        reject(e);
+      }
+    });
```

Comment:
This is more old code that could be simplified to e.g.
```javascript
return new Promise(function() {
  const value = obj[prop];
  let result;
  if (typeof value === 'function') {
    result = value.apply(obj, args);
  } else {
    result = value;
  }
  return result;
});
```

Reply:
I tried doing this, but I keep on getting unit test failures with this change:

```
TEST-UNEXPECTED-FAIL | creates pdf doc from URL | Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
```

I don't really see why both would be different, but apparantly not explicitly rejecting the promise causes different behaviour?!

Reply:
That's really strange; anyway I'm sorry for wasting your time with bad advice!

Reply:
Jonas' remark on simplicity is correct, in the sense that you don't have to explicitly catch errors and reject errors (that is the default behavior). However, you have to call `resolve(result)` instead of `return result;` in the end.

We don't use async/await anywhere in the codebase (yet?), but if we were, then the code could also be simplified to (without `try-catch` and without the explicit `Promise` constructor):

```javascript
async ensure(obj, prop, args) {
  const value = obj[prop];
  if (typeof value === 'function') {
    return value.apply(obj, args);
  }
  return value;
}
```

Reply:
D'oh, that's most likely what I missed... I'll give this another try when I'll address the other comments. I have been reading a bit about `async`/`await` and it does look really nice. Since we've not used it yet, I'll stick with regular promises for now, but it's something we may consider for the future (I think Brendan mentioned it some time ago as well).

Reply:
Regarding using `async`/`await`: Currently Babel doesn't have great support for this, if you want the output to be IE11 compatible (which we unfortunately do). There's the upstream issue https://github.com/babel/babel/issues/8121 and in-progress PR https://github.com/babel/babel/pull/7076 which looks relevant here, and are probably worth following for updates.